File: src/commands/upgrade.ts
import * as p from '@clack/prompts';
import chalk from 'chalk';
import { join } from 'node:path';
import { exists, readText, writeText, ensureDir } from '../utils/fs.js';
import { readHashes, writeHashes, hashContent, hashFile, type HashManifest } from '../utils/hash.js';
import { readYaml } from '../utils/yaml.js';
import { render, type TemplateContext } from '../render/engine.js';
import { mergeForgeYaml } from '../render/merge.js';
import { resolveTemplatePath } from '../utils/fs.js';
import { detect } from '../detect/index.js';
import { dirname } from 'node:path';

interface UpgradeOptions {
	force?: boolean;
}

// Files that are ALWAYS overwritten (tool-owned)
const ALWAYS_OVERWRITE = ['.forge/context/stack.md'];

// Files that are NEVER touched (user-owned)
const NEVER_TOUCH = ['.forge/context/project.md'];

export async function upgrade(options: UpgradeOptions): Promise<void> {
	const cwd = process.cwd();

	p.intro(chalk.bold('forge upgrade'));

	if (!(await exists(join(cwd, 'forge.yaml')))) {
		p.log.error('No forge.yaml found. Run `forge init` first.');
		process.exit(1);
	}

	const hashes = await readHashes(cwd);
	const currentVersion = hashes.version;
	const newVersion = '0.1.0'; // TODO: read from package.json

	p.log.info(`Current: v${currentVersion} → Available: v${newVersion}`);

	const config = await readYaml<Record<string, unknown>>(join(cwd, 'forge.yaml'));
	const project = config.project as Record<string, string>;
	const preset = project?.preset ?? 'sveltekit-ts';

	// Rebuild template context
	const detected = await detect(cwd);
	const ctx = buildUpgradeContext(config, detected, preset);

	let updated = 0;
	let skipped = 0;
	let prompted = 0;
	const newHashes: HashManifest = { version: newVersion, files: {} };

	// Get all template files that should exist
	for (const [relativePath, installedHash] of Object.entries(hashes.files)) {
		// Skip addon files — they're managed separately
		if (relativePath.startsWith('.forge/addons/')) continue;

		// Never touch user-owned files
		if (NEVER_TOUCH.includes(relativePath)) {
			p.log.message(`  ${chalk.dim('⊘')} ${relativePath} ${chalk.dim('— user-owned, skipped')}`);
			skipped++;
			// Keep the existing hash
			newHashes.files[relativePath] = installedHash;
			continue;
		}

		const filePath = join(cwd, relativePath);
		if (!(await exists(filePath))) {
			// File was deleted by user — skip
			skipped++;
			continue;
		}

		// Try to find and render the new template
		const templatePath = findTemplatePath(relativePath, preset);
		if (!templatePath) {
			newHashes.files[relativePath] = installedHash;
			continue;
		}

		let newContent: string;
		try {
			const template = await readText(resolveTemplatePath(templatePath));
			newContent = render(template, ctx);
		} catch {
			newHashes.files[relativePath] = installedHash;
			continue;
		}

		const newContentHash = hashContent(newContent);

		// Always overwrite tool-owned files
		if (ALWAYS_OVERWRITE.includes(relativePath)) {
			await writeText(filePath, newContent);
			newHashes.files[relativePath] = newContentHash;
			p.log.success(`${relativePath} ${chalk.dim('— tool-owned, updated')}`);
			updated++;
			continue;
		}

		// Check if user has modified the file
		const currentHash = await hashFile(filePath);
		const userModified = currentHash !== installedHash;

		if (!userModified || options.force) {
			// User hasn't modified — safe to overwrite
			await writeText(filePath, newContent);
			newHashes.files[relativePath] = newContentHash;
			if (newContentHash !== installedHash) {
				p.log.success(`${relativePath} ${chalk.dim('— updated')}`);
				updated++;
			}
		} else {
			// User modified — ask
			const action = await p.select({
				message: `${relativePath} — you modified this. What to do?`,
				options: [
					{ value: 'skip', label: 'Skip (keep your version)' },
					{ value: 'overwrite', label: 'Overwrite (use new version)' },
				],
			});

			if (p.isCancel(action)) {
				p.cancel('Upgrade cancelled.');
				process.exit(0);
			}

			if (action === 'overwrite') {
				await writeText(filePath, newContent);
				newHashes.files[relativePath] = newContentHash;
				updated++;
			} else {
				newHashes.files[relativePath] = currentHash;
				skipped++;
			}
			prompted++;
		}
	}

	// Merge forge.yaml — add new fields, preserve existing
	try {
		const templatePath = resolveTemplatePath('core', 'forge.yaml.hbs');
		const template = await readText(templatePath);
		const newForgeYaml = render(template, ctx);
		const existingForgeYaml = await readText(join(cwd, 'forge.yaml'));
		const merged = mergeForgeYaml(existingForgeYaml, newForgeYaml);
		if (merged !== existingForgeYaml) {
			await writeText(join(cwd, 'forge.yaml'), merged);
			p.log.success('forge.yaml — merged new fields');
			updated++;
		}
	} catch {
		// Merge failed — skip
	}

	await writeHashes(cwd, newHashes);

	p.outro(
		`Updated to v${newVersion}. ${chalk.green(`${updated} updated`)}, ${chalk.dim(`${skipped} skipped`)}${prompted > 0 ? `, ${chalk.yellow(`${prompted} prompted`)}` : ''}`,
	);
}

function findTemplatePath(relativePath: string, preset: string): string | null {
	// Map output paths back to template paths
	const mappings: Record<string, string> = {
		'forge.yaml': 'core/forge.yaml.hbs',
		'CLAUDE.md': 'core/CLAUDE.md.hbs',
		'.claude/settings.json': 'core/settings.json.hbs',
		'.claude/skills/deliver/SKILL.md': 'core/skill-deliver.md.hbs',
		'.forge/pipeline/orchestrator.sh': 'core/pipeline/orchestrator.sh.hbs',
		'.forge/pipeline/intake.sh': 'core/pipeline/intake.sh.hbs',
		'.forge/pipeline/classify.sh': 'core/pipeline/classify.sh.hbs',
		'.forge/pipeline/decompose.md': 'core/pipeline/decompose.md.hbs',
		'.forge/pipeline/execute.md': 'core/pipeline/execute.md.hbs',
		'.forge/pipeline/verify.sh': 'core/pipeline/verify.sh.hbs',
		'.forge/pipeline/deliver.sh': 'core/pipeline/deliver.sh.hbs',
		'.forge/agents/architect.md': 'core/agents/architect.md.hbs',
		'.forge/agents/quality.md': 'core/agents/quality.md.hbs',
		'.forge/agents/security.md': 'core/agents/security.md.hbs',
		'.forge/agents/frontend.md': 'core/agents/frontend.md.hbs',
		'.forge/agents/backend.md': 'core/agents/backend.md.hbs',
		'.forge/context/stack.md': `presets/${preset}/stack.md.hbs`,
		'.forge/context/project.md': 'core/context/project.md.hbs',
		'.forge/hooks/pre-edit.sh': 'core/hooks/pre-edit.sh.hbs',
		'.forge/hooks/post-edit.sh': 'core/hooks/post-edit.sh.hbs',
		'.forge/hooks/session-start.sh': 'core/hooks/session-start.sh.hbs',
	};

	return mappings[relativePath] ?? null;
}

function buildUpgradeContext(
	config: Record<string, unknown>,
	detected: import('../detect/index.js').DetectedStack,
	preset: string,
): TemplateContext {
	const project = config.project as Record<string, string>;
	const commands = config.commands as Record<string, string>;
	const agents = (config.agents as string[]) ?? [];
	const pipeline = config.pipeline as Record<string, unknown> ?? {};
	const isFrontendPreset = ['sveltekit-ts', 'react-next-ts', 'vue-nuxt-ts'].includes(preset);

	return {
		project: {
			name: project?.name ?? 'my-app',
			preset,
		},
		commands: commands ?? {},
		agents,
		has_frontend: isFrontendPreset,
		has_format: Boolean(commands?.format),
		auto_pr: pipeline?.auto_pr ?? true,
		detected: {
			language: detected.language,
			framework: detected.framework,
			features: detected.features,
		},
		preset,
		is_sveltekit: preset === 'sveltekit-ts',
		is_nextjs: preset === 'react-next-ts',
		is_fastapi: preset === 'python-fastapi',
		is_go: preset === 'go',
	};
}

File: cmd/upgrade.go
package cmd

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"

	"github.com/samahlstrom/forge-cli/internal/detect"
	"github.com/samahlstrom/forge-cli/internal/render"
	"github.com/samahlstrom/forge-cli/internal/resolve"
	"github.com/samahlstrom/forge-cli/internal/static"
	"github.com/samahlstrom/forge-cli/internal/ui"
	"github.com/samahlstrom/forge-cli/internal/util"

	"github.com/spf13/cobra"
)

var (
	upgradeForce  bool
	upgradeGlobal bool
)

func init() {
	upgradeCmd := &cobra.Command{
		Use:   "upgrade",
		Short: "Upgrade forge — global library and/or project files",
		Long: `Upgrades forge files to the latest version.

By default, upgrades both ~/.forge/ (global library) and the current project.
Use --global to upgrade only the global library (no project needed).`,
		RunE: runUpgrade,
	}
	upgradeCmd.Flags().BoolVar(&upgradeForce, "force", false, "Overwrite all files without prompting")
	upgradeCmd.Flags().BoolVar(&upgradeGlobal, "global", false, "Upgrade only the global library (~/.forge/)")
	rootCmd.AddCommand(upgradeCmd)
}

// Files that are ALWAYS overwritten (tool-owned).
var alwaysOverwrite = map[string]bool{
	".forge/context/stack.md":  true,
	".claude/settings.json":    true,
}

// Files that are NEVER touched (user-owned).
var neverTouch = map[string]bool{
	".forge/context/project.md": true,
}

func runUpgrade(cmd *cobra.Command, args []string) error {
	cwd, _ := os.Getwd()

	// Guard: prevent upgrading forge inside its own source repo
	if isForgeSelfRepo(cwd) {
		ui.Log.Error("Cannot run `forge upgrade` inside the forge-cli source repository.")
		return fmt.Errorf("self-upgrade blocked")
	}

	ui.Intro(ui.Bold("forge upgrade"))

	// Always upgrade global library
	ui.Log.Info(fmt.Sprintf("Upgrading global library at %s", resolve.ForgeHome()))
	globalCount, err := bootstrapGlobal(resolve.ForgeHome(), upgradeForce)
	if err != nil {
		return fmt.Errorf("global upgrade failed: %w", err)
	}
	ui.Log.Success(fmt.Sprintf("Global library: %d files updated", globalCount))

	// If --global only, stop here
	if upgradeGlobal {
		ui.Outro(fmt.Sprintf("Global library upgraded to v%s.", static.Version))
		return nil
	}

	// Project upgrade — only project-local files
	if !util.Exists(filepath.Join(cwd, "forge.yaml")) {
		ui.Log.Step("No forge.yaml found in current directory — skipping project upgrade.")
		ui.Outro(fmt.Sprintf("Global library upgraded to v%s.", static.Version))
		return nil
	}

	hashes, _ := util.ReadHashes(cwd)
	newVersion := static.Version

	var config map[string]any
	if err := util.ReadYAML(filepath.Join(cwd, "forge.yaml"), &config); err != nil {
		return err
	}
	project, _ := config["project"].(map[string]any)
	preset, _ := project["preset"].(string)
	if preset == "" {
		preset = "sveltekit-ts"
	}

	detected := detect.Detect(cwd)
	ctx := buildUpgradeContext(config, detected, preset)

	updated := 0
	skipped := 0
	newHashes := util.HashManifest{Version: newVersion, Files: map[string]string{}}

	for relPath, installedHash := range hashes.Files {
		// Only upgrade project-local files
		if !isProjectFile(relPath) {
			continue
		}

		// Never touch user-owned files
		if neverTouch[relPath] {
			newHashes.Files[relPath] = installedHash
			skipped++
			continue
		}

		filePath := filepath.Join(cwd, relPath)
		if !util.Exists(filePath) {
			skipped++
			continue
		}

		templatePath := findTemplatePath(relPath, preset)
		if templatePath == "" {
			newHashes.Files[relPath] = installedHash
			continue
		}

		templateContent, err := fs.ReadFile(static.TemplatesFS, filepath.Join("templates", templatePath))
		if err != nil {
			newHashes.Files[relPath] = installedHash
			continue
		}
		newContent := render.Render(string(templateContent), ctx)
		newContentHash := util.HashContent(newContent)

		// Always overwrite tool-owned files
		if alwaysOverwrite[relPath] {
			_ = util.WriteText(filePath, newContent)
			newHashes.Files[relPath] = newContentHash
			ui.Log.Success(fmt.Sprintf("%s %s", relPath, ui.Dim("— updated")))
			updated++
			continue
		}

		// Check if user modified the file
		currentHash, _ := util.HashFile(filePath)
		userModified := currentHash != installedHash

		if !userModified || upgradeForce {
			_ = util.WriteText(filePath, newContent)
			newHashes.Files[relPath] = newContentHash
			if newContentHash != installedHash {
				ui.Log.Success(fmt.Sprintf("%s %s", relPath, ui.Dim("— updated")))
				updated++
			}
		} else {
			action, cancelled := ui.Select(
				fmt.Sprintf("%s — you modified this. What to do?", relPath),
				[]ui.SelectOption{
					{Value: "skip", Label: "Skip (keep your version)"},
					{Value: "overwrite", Label: "Overwrite (use new version)"},
				},
			)
			if cancelled {
				ui.Cancel("Upgrade cancelled.")
				os.Exit(0)
			}
			if action == "overwrite" {
				_ = util.WriteText(filePath, newContent)
				newHashes.Files[relPath] = newContentHash
				updated++
			} else {
				newHashes.Files[relPath] = currentHash
				skipped++
			}
		}
	}

	// Merge forge.yaml (add new keys only)
	templateContent, err := fs.ReadFile(static.TemplatesFS, filepath.Join("templates", "core", "forge.yaml.hbs"))
	if err == nil {
		newForgeYAML := render.Render(string(templateContent), ctx)
		existingForgeYAML, _ := util.ReadText(filepath.Join(cwd, "forge.yaml"))
		merged, mergeErr := render.MergeForgeYAML(existingForgeYAML, newForgeYAML)
		if mergeErr == nil && merged != existingForgeYAML {
			_ = util.WriteText(filepath.Join(cwd, "forge.yaml"), merged)
			ui.Log.Success("forge.yaml — merged new fields")
			updated++
		}
	}

	// Clean up deprecated per-project files (agents, pipeline, hooks now live in ~/.forge/)
	deprecatedProjectFiles := []string{
		".forge/pipeline/classify.sh",
		".forge/pipeline/orchestrator.sh",
		".forge/pipeline/helpers.sh",
		".forge/pipeline/intake.sh",
		".forge/pipeline/classify.md",
		".forge/pipeline/review-plan.md",
		".forge/pipeline/verify.sh",
		".forge/pipeline/deliver.sh",
		".forge/pipeline/browser-smoke.sh",
		".forge/hooks/pre-edit.sh",
		".forge/hooks/post-edit.sh",
		".forge/hooks/session-start.sh",
	}
	for _, dep := range deprecatedProjectFiles {
		depPath := filepath.Join(cwd, dep)
		if util.Exists(depPath) {
			_ = os.Remove(depPath)
			ui.Log.Step(fmt.Sprintf("%s %s", dep, ui.Dim("— moved to ~/.forge/, removed local copy")))
		}
	}

	// Clean up deprecated agent files (now in ~/.forge/agents/)
	agentsDir := filepath.Join(cwd, ".forge", "agents")
	if util.Exists(agentsDir) {
		entries, _ := os.ReadDir(agentsDir)
		if len(entries) == 0 {
			_ = os.Remove(agentsDir)
		}
	}

	// Clean up deprecated directories
	for _, dir := range []string{".claude/skills/deliver", ".forge/hooks", ".forge/addons", ".forge/state"} {
		dirPath := filepath.Join(cwd, dir)
		if util.Exists(dirPath) {
			entries, _ := os.ReadDir(dirPath)
			if len(entries) == 0 {
				_ = os.Remove(dirPath)
			}
		}
	}

	_ = util.WriteHashes(cwd, newHashes)

	ui.Outro(fmt.Sprintf("Upgraded to v%s. %s project files updated, %s skipped.",
		newVersion,
		ui.Green(fmt.Sprintf("%d", updated)),
		ui.Dim(fmt.Sprintf("%d", skipped)),
	))
	return nil
}

// isProjectFile returns true if the file is a project-local file (not a global library file).
func isProjectFile(relPath string) bool {
	projectFiles := map[string]bool{
		"forge.yaml":              true,
		"CLAUDE.md":               true,
		".claude/settings.json":   true,
		".forge/context/stack.md": true,
		".forge/context/project.md": true,
	}
	return projectFiles[relPath]
}

func findTemplatePath(relPath, preset string) string {
	mappings := map[string]string{
		"forge.yaml":              "core/forge.yaml.hbs",
		"CLAUDE.md":               "core/CLAUDE.md.hbs",
		".claude/settings.json":   "core/settings.json.hbs",
		".forge/context/stack.md": fmt.Sprintf("presets/%s/stack.md.hbs", preset),
		".forge/context/project.md": "core/context/project.md.hbs",
	}
	return mappings[relPath]
}

func buildUpgradeContext(config map[string]any, detected detect.DetectedStack, preset string) render.Ctx {
	project, _ := config["project"].(map[string]any)
	commands, _ := config["commands"].(map[string]any)
	agents := toStringSlice(config["agents"])
	pipeline, _ := config["pipeline"].(map[string]any)
	isFrontend := preset == "sveltekit-ts" || preset == "react-next-ts" || preset == "vue-nuxt-ts"

	name, _ := project["name"].(string)
	if name == "" {
		name = "my-app"
	}

	autoPR := true
	if pipeline != nil {
		if v, ok := pipeline["auto_pr"].(bool); ok {
			autoPR = v
		}
	}

	cmdMap := map[string]any{}
	if commands != nil {
		cmdMap = commands
	}

	return render.Ctx{
		"project": map[string]any{
			"name":   name,
			"preset": preset,
		},
		"commands":     cmdMap,
		"agents":       agents,
		"has_frontend": isFrontend,
		"has_format":   commands != nil && commands["format"] != nil && commands["format"] != "",
		"auto_pr":      autoPR,
		"detected": map[string]any{
			"language":  detected.Language,
			"framework": detected.Framework,
			"features":  map[string]any{"git": detected.Features.Git, "ci": detected.Features.CI, "docker": detected.Features.Docker},
		},
		"preset":     preset,
		"is_sveltekit": preset == "sveltekit-ts",
		"is_nextjs":    preset == "react-next-ts",
		"is_fastapi":   preset == "python-fastapi",
		"is_go":        preset == "go",
		"stackFile":    ".forge/context/stack.md",
		"projectFile":  ".forge/context/project.md",
	}
}
